teps/0033-tekton-feature-gates.md

```diff
@@ -145,6 +160,393 @@ alpha features. This is not a requirement for Tekton, because even in case of
+  on an alpha feature
+    * In the future we may try to communicate this to the user somehow, e.g. through something in the
+      run status (probably through logs at the very least)
+* Tools that integrate with Tekton may have a hard time telling the difference between alpha
+  fields and beta/v1, e.g. a vscode integration with auto completion
+    * Hopefully we can mitigate this with documentation; it will be up to the integrating tools
+      to decide how to handle this
+* These two fields would toggle alpha and beta fields across all CRDs; another option would be
+  to have a flag per version per CRD
+
+#### Docs
+
+We will maintain documentation showing the status of each field:
+
+* We will create a table of fields that are considered alpha or beta (which will link to docs
+  on each feature)
+* In the docs for each CRD, we will have a section for alpha fields and beta fields
+
+### Changes in behavior
+
+Changes in behavior are features that are not controlled by CRD fields, for example if we
+wanted to change finally tasks such that they always run when a PipelineRun is cancelled.
+
+When we need to make behavior changes like this, each will get its own
+[feature flag](https://github.com/tektoncd/pipeline/blob/master/docs/install.md#customizing-the-pipelines-controller-behavior),
+like existing feature flags such as `disable-affinity-assistant` and `disable-home-env-overwrite`.
+
+_We should avoid adding features like this if possible since each new feature
+complicates the combination of settings we'll need to test in order to feel
+confident in our automated testing, and because promoting these changes from
+alpha will be a backwards incompatible change._ 
+
+Pros:
+* It will be clear to users whether they have opted into this functionality or not
+
+Cons:
+* Flags added this way can only be removed according to our deprecation policy and will need to
+  be supported for as long as the corresponding version is supported (e.g. today we support v1alpha1
+  and v1beta1 so flags of this kind that we added for v1beta1 functionality will stick around until
+  we completely remove support for v1beta1)
+
+### Promotion to beta and beyond
+
+It will be at the discretion of [the pipelines OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS_ALIASES)
+to decide when a feature is ready to move from alpha to beta and finally to v1.
+
+This will likely require the feature being available for at least one release
+if not several releases (and maybe use in [dogfooding](https://github.com/tektoncd/plumbing/blob/main/docs/dogfooding.md))
+to be able to collect and handle user feedback.
+
+* When an alpha API field is ready to be declared beta:
+    * If the underlying CRD has a version of v1, the webhook admission controller
+      will be updated to require the `enable-beta-api-fields` flag instead of the
+      `enable-alpha-api-fields` flag in order to use the field.
```

Comment:
Echoing @afrittoli 's comment above - I don't think users should be able to enable alpha fields while disabling beta ones because it seems reasonable to me for alpha fields to augment functionality introduced in beta fields. Disabling beta while enabling alpha would make that potentially quite hairy?  Suggested change:

```
If the underlying CRD has a version of v1, the webhook admission controller
will be updated to support the feature under the `enable-beta-api-fields` flag
in addition to its existing support under the `enable-alpha-api-fields` flag.
```